crypto/lms/lm_ots_verify.c

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include "crypto/lms_sig.h"
+#include "crypto/lms_util.h"
+
+static int lm_ots_compute_pubkey_final(EVP_MD_CTX *ctx, EVP_MD_CTX *ctxIq,
+                                       const LM_OTS_SIG *sig, unsigned char *Kc);
+
+/**
+ * @brief OTS Signature verification.
+ *
+ * See RFC 8554 Section 4.6 Signature Verification
+ * Algorithm 4b: Compute a Public Key Candidate |Kc| from a signature, message
+ * and public key parameters.
+ *
+ * @param ctx A temporary working EVP_MD_CTX object.
+ * @param ctxIq A EVP_MD_CTX object that has been initialised with a digest,
+ *              that returns a non finalized value of H(I || q)
+ * @param sig An LM_OTS_SIG object that contains C and y
+ * @param pub The public key LM_OTS_PARAMS
+ * @param Id A 16 byte indentifier (I) associated with a LMS tree
+ * @param q The leaf index of the LMS tree.
+ * @param msg A message to verify
+ * @param msglen The size of |msg|
+ * @param Kc The computed public key candidate. It is assumed the size is n.
+ * @returns 1 on success, or 0 otherwise.
+ */
+int ossl_lm_ots_compute_pubkey(EVP_MD_CTX *ctx, EVP_MD_CTX *ctxIq,
+                               const LM_OTS_SIG *sig, const LM_OTS_PARAMS *pub,
+                               const unsigned char *Id, uint32_t q,
+                               const unsigned char *msg, size_t msglen,
+                               unsigned char *Kc)
+{
+    int ret = 0;
+    unsigned char qbuf[LMS_SIZE_q];
+
+    if (sig->params != pub)
+        return 0;
+
+    U32STR(qbuf, q);
+
+    if (!EVP_DigestUpdate(ctxIq, Id, LMS_SIZE_I)
+            || !EVP_DigestUpdate(ctxIq, qbuf, sizeof(qbuf))
+            || !EVP_MD_CTX_copy_ex(ctx, ctxIq)
+            /* Q = H(I || u32str(q) || u16str(D_MESG) || C || msg) */
+            || !EVP_DigestUpdate(ctx, &OSSL_LMS_D_MESG, sizeof(OSSL_LMS_D_MESG))
```

Comment:
This doesn't appear to be endian safe.

Is this not caught by CI?

(repeat this comment for other instances)

Reply:
The value is `0x8181` so endian issues aren't a concern.
Still it is better fixed.

Reply:
I've also updated the other instances of these.  They are all repeated bytes so it's a nop.